debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresPendingLsnStore.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.debezium.connector.postgresql;
+
+import java.io.StringWriter;
+import java.util.Collection;
+import java.util.Enumeration;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.kafka.connect.source.SourceRecord;
+import org.postgresql.replication.LogSequenceNumber;
+
+public class PostgresPendingLsnStore {
+    private final ConcurrentHashMap<Long, Integer> lsnsInProgress = new ConcurrentHashMap<>();
+
+    public void recordPolledLsn(SourceRecord record) {
+        Long lsn = getLsnFromRecord(record);
+        if (lsn == null) {
+            return;
+        }
+
+        lsnsInProgress.compute(lsn, (_lsn, count) -> {
+            if (count == null) {
+                return 1;
+            } else {
+                return count + 1;
+            }
+        });
+    }
+
+    public void recordPolledLsns(Collection<SourceRecord> records) {
+        records.forEach(this::recordPolledLsn);
+    }
+
+    public void recordProcessedLsn(SourceRecord record) throws IllegalStateException {
+        Long lsn = getLsnFromRecord(record);
+        if (lsn == null) {
+            return;
+        }
+
+        lsnsInProgress.compute(lsn, (_lsn, count) -> {
+            if (count == null || count < 0) {
+                throw new IllegalStateException("Attempted to record processed LSN when it hasn't been polled");
+            } else if (count == 1) {
+                return null;
+            } else {
+                return count - 1;
+            }
+        });
+    }
+
+    public Long getEarliestUnprocessedLsn() {
+        // If there are no elements, return null
+        Enumeration<Long> lsnIterator = lsnsInProgress.keys();
+        if (!lsnIterator.hasMoreElements()) {
+            return null;
+        }
+
+        // Otherwise, iterate through each lsn to find the smallest
+        Long earliestLsn = lsnIterator.nextElement();
+        while (lsnIterator.hasMoreElements()) {
+            Long lsn = lsnIterator.nextElement();
+            if (lsn < earliestLsn) {
```

Comment:
Now that I'm looking at the code, I realized this part still isn't quite right... What we end up finding here is the smallest LSN that still has a record that hasn't been committed, meaning we still have the chance for data loss. I think we need to either:

1. Change the `recordProcessedLsn` to not remove items from the map when their count goes to 0, find the correct value here (largest LSN where all prior LSNs have zero count), and then go through and remove everything lower than that LSN

2. When we remove an LSN in `recordProcessedLsn`, we stick it in another data structure. In the code here, we still find the smallest LSN in the map, but then find the largest LSN in the other data structure less than that value, and commit that LSN (and cleanup the second data structure).

3. Switch to the concurrent skip list implementation, with O(1) min value lookup, then each time we remove an LSN because its count is 0, check if it's less than the list's min value, if so we know we can commit that LSN.

4. Something smarter I'm not thinking of

I think I'd be in favor of 2?

Reply:
We chatted offline on Slack, but to keep a record here on the PR: This probably works fine as-is, because it looks like the LSN stored in the replication slot is where Postgres should pick up _from_, not pick up _after_. We found some other documentation to support this, but here it is in Postgres: https://www.postgresql.org/docs/11/view-pg-replication-slots.html

The documentation is still a _little_ unclear, so I'm going to try a test locally to make sure we understand the behavior right. I'll produce records at a slow rate, keep a close eye on the logs and when an LSN gets flushed, and then try restarting to make sure we pick up from the correct spot.

Reply:
Revisiting this: After a lot of investigation, it turns out what I was hoping for above ^ doesn't quite work out. Part of the problem is that there are actually multiple WAL items even for individual inserts (`INSERT`, `COMMIT`, and `RUNNING_XACTS`), and the LSN we assign to a record is the LSN of the _last_ of those log items. If the stream tries to pick up again from that position, it'll be past the log item with the actual record payload.

```
rmgr: Heap        len (rec/tot):     61/    61, tx:        555, lsn: 0/016B1C80, prev 0/016B1BB0, desc: INSERT+INIT off 1, blkref #0: rel 1663/16384/16387 blk 0
rmgr: Transaction len (rec/tot):     46/    46, tx:        555, lsn: 0/016B1CC0, prev 0/016B1C80, desc: COMMIT 2020-03-24 22:22:11.621616 UTC
rmgr: Standby     len (rec/tot):     50/    50, tx:          0, lsn: 0/016B1CF0, prev 0/016B1CC0, desc: RUNNING_XACTS nextXid 556 latestCompletedXid 555 oldestRunningXid 556
rmgr: Heap        len (rec/tot):     61/    61, tx:        556, lsn: 0/016B1D28, prev 0/016B1CF0, desc: INSERT off 2, blkref #0: rel 1663/16384/16387 blk 0
rmgr: Transaction len (rec/tot):     46/    46, tx:        556, lsn: 0/016B1D68, prev 0/016B1D28, desc: COMMIT 2020-03-24 22:22:21.754083 UTC
rmgr: Standby     len (rec/tot):     50/    50, tx:          0, lsn: 0/016B1D98, prev 0/016B1D68, desc: RUNNING_XACTS nextXid 557 latestCompletedXid 556 oldestRunningXid 557
```

Reply:
I agree with you that 2 is probably the best option. Polling N items remains O(N), and processing a single item remains O(1). Finding the right one on commit isn't even too expensive (O(X) to find the smallest uncommitted, O(Y) to find the largest comitted that's less than the uncomitted, and then another O(Y) pass to remove the appropriate things). All linear.

I imagine 1 is more expensive, as the fastest way to find the largest LSN with a count of 0 that has all counts below it also as 0 probably involves doing a sort first. 3 is probably expensive as well, since inserting things in sorted order is usually O(log(N)) or so.